apps/evm_family/evm_txn_helpers.c

```diff
@@ -173,38 +174,43 @@ static uint64_t get_decode_length(const uint8_t *seq,
   return item_bytes_len;
 }
 
-static PAYLOAD_STATUS eth_decode_txn_payload(
-    const evm_unsigned_txn *evm_utxn_ptr,
-    txn_metadata *metadata_ptr) {
+static PAYLOAD_STATUS evm_decode_txn_payload(
+    const evm_sign_txn_initiate_request_t *init_info,
+    const evm_unsigned_txn *utxn_ptr,
+    const erc20_contracts_t **contract) {
   PAYLOAD_STATUS result = PAYLOAD_ABSENT;
-  eth_is_token_whitelisted = false;
-  if (evm_utxn_ptr->payload_size > 0) {
-    if ((evm_utxn_ptr->payload_size >= 4) &&
-        (U32_READ_BE_ARRAY(evm_utxn_ptr->payload) == TRANSFER_FUNC_SIGNATURE) &&
-        (metadata_ptr->is_token_transfer) &&
-        (metadata_ptr->network_chain_id == ETHEREUM_MAINNET_CHAIN)) {
-      for (int16_t i = 0; i < 0; i++) {
-        if (strncmp(metadata_ptr->token_name,
+  evm_is_token_whitelisted = false;
+
+  if (utxn_ptr->payload_size > 0) {
+    if ((utxn_ptr->payload_size >= 4) &&
+        (U32_READ_BE_ARRAY(utxn_ptr->payload) == TRANSFER_FUNC_SIGNATURE) &&
+        (init_info->is_token_transfer) &&
+        (init_info->chain_id == ETHEREUM_MAINNET_CHAIN)) {
+      for (int16_t i = 0; i < g_evm_app->whitelist_count; i++) {
+        if (strncmp(init_info->token_symbol,
                     g_evm_app->whitelisted_contracts[i].symbol,
                     ETHEREUM_TOKEN_SYMBOL_LENGTH) == 0) {
-          metadata_ptr->eth_val_decimal[0] =
-              g_evm_app->whitelisted_contracts[i].decimal;
-          eth_is_token_whitelisted = true;
-          result = (memcmp(evm_utxn_ptr->to_address,
+          evm_is_token_whitelisted = true;
+          result = (memcmp(utxn_ptr->to_address,
                            g_evm_app->whitelisted_contracts[i].address,
                            EVM_ADDRESS_LENGTH) == 0)
                        ? PAYLOAD_WHITELISTED
                        : PAYLOAD_CONTRACT_INVALID;
+
+          if (PAYLOAD_WHITELISTED == result) {
+            *contract = &g_evm_app->whitelisted_contracts[i];
+          }
           break;
         }
       }
     }
-    if (!eth_is_token_whitelisted)
-      result = (ETH_ExtractArguments(evm_utxn_ptr->payload,
-                                     evm_utxn_ptr->payload_size) ==
-                ETH_UTXN_ABI_DECODE_OK)
-                   ? PAYLOAD_WHITELISTED
-                   : PAYLOAD_SIGNATURE_NOT_WHITELISTED;
+    if (!evm_is_token_whitelisted) {
+      result =
+          (ETH_ExtractArguments(utxn_ptr->payload, utxn_ptr->payload_size) ==
+           ETH_UTXN_ABI_DECODE_OK)
+              ? PAYLOAD_WHITELISTED
+              : PAYLOAD_SIGNATURE_NOT_WHITELISTED;
+    }
```

Comment:
Let's add a todo here for detecting clear sign or blind sign transaction using the function tag value along with function signature.

Reply:
added